bundle/Routing/SiteaccessResolver.php

```diff
@@ -0,0 +1,388 @@
+                    $availableLanguageSet,
+                    $alwaysAvailable
+                );
+
+                if ($maybeSiteaccess !== null) {
+                    return $maybeSiteaccess;
+                }
+            }
+        }
+
+        if ($alwaysAvailable) {
+            return reset($siteaccesses);
+        }
+
+        return $currentSiteaccess;
+    }
+
+    private function matchTranslationSiteaccess(
+        ?string $siteaccess,
+        array $currentPrioritizedLanguages,
+        array $availableLanguageSet,
+        bool $alwaysAvailable
+    ): ?string {
+        if ($siteaccess === null) {
+            return null;
+        }
+
+        foreach ($currentPrioritizedLanguages as $language) {
+            if (!array_key_exists($language, $availableLanguageSet)) {
+                continue;
+            }
+
+            $maybeTranslationSiteaccess = $this->matchTranslationSiteaccessByLanguage($siteaccess, $language);
+
+            if ($maybeTranslationSiteaccess !== null) {
+                return $maybeTranslationSiteaccess;
+            }
+        }
+
+        if ($alwaysAvailable) {
+            return $siteaccess;
+        }
+
+        if ($this->isSomeLanguageAllowed($siteaccess, $availableLanguageSet)) {
+            return $siteaccess;
+        }
+
+        return null;
+    }
+
+    private function isSomeLanguageAllowed(string $siteaccess, array $availableLanguageSet): bool
+    {
+        $prioritizedLanguages = $this->getPrioritizedLanguages($siteaccess);
+
+        foreach ($prioritizedLanguages as $language) {
+            if (array_key_exists($language, $availableLanguageSet)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private function matchTranslationSiteaccessByLanguage(string $siteaccess, string $language): ?string
+    {
+        $translationSiteaccesses = $this->getTranslationSiteaccesses($siteaccess);
+
+        foreach ($translationSiteaccesses as $translationSiteaccess) {
+            if ($this->isSiteaccessExcluded($translationSiteaccess)) {
+                continue;
+            }
+
+            $primaryLanguage = $this->getPrimaryLanguage($translationSiteaccess);
+
+            if ($language === $primaryLanguage) {
+                return $translationSiteaccess;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param string[] $siteaccesses
+     *
+     * @return string[]
+     */
+    private function matchSiteaccessesByPrimaryLanguage(array $siteaccesses, string $language): array
+    {
+        $matched = [];
+
+        foreach ($siteaccesses as $siteaccess) {
+            $primaryLanguage = $this->getPrimaryLanguage($siteaccess);
+
+            if ($language === $primaryLanguage) {
+                $matched[] = $siteaccess;
+            }
+        }
+
+        return $matched;
+    }
+
+    /**
+     * @param string[] $siteaccesses
+     */
+    private function matchSiteaccessesByBestPrioritizedLanguage(
```

Comment:
Matching translation siteaccess can be moved inside this to avoid unnecessary recursion.